### PR TITLE
Feature/001 add new location

### PR DESCRIPTION
**Pull request by MichalKulygin**
feature011_add_new_location finished

### CODE REVIEW

File: src/main/java/com/mkgn/weatherservice/location/Cardinals.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package com.mkgn.weatherservice.location;
+
+public enum Cardinals {
```
Comment (klezner) [inline src/main/java/com/mkgn/weatherservice/location/Cardinals.java:3]: Cardinals are unnecessary

Comment (klezner) [inline src/main/java/com/mkgn/weatherservice/location/Cardinals.java:3]: Cardinals are unnecessary

File: src/main/java/com/mkgn/weatherservice/location/NewLocationResponse.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+@AllArgsConstructor
+@Builder
+class NewLocationResponse {
+    private String id;
+    private String city;
+    private String region;
+    private String country;
+    private Double latitude;
+    private Double longitude;
+    private String latitudeCardinal;
```
Comment (klezner) [inline src/main/java/com/mkgn/weatherservice/location/NewLocationResponse.java:19]: Unnecessary field

File: src/main/java/com/mkgn/weatherservice/location/NewLocationResponse.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    private String longitudeCardinal;
```
Comment (klezner) [inline src/main/java/com/mkgn/weatherservice/location/NewLocationResponse.java:20]: Unnecessary field

File: src/main/java/com/mkgn/weatherservice/location/LocationMapper.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+                .region(location.getRegion())
+                .country(location.getCountry())
+                .latitude(location.getLatitude())
+                .longitude(location.getLongitude())
+                .latitudeCardinal(latitudeCardinal(location.getLatitude()))
+                .longitudeCardinal(longitudeCardinal(location.getLongitude()))
+                .build();
+    }
+
+    private String longitudeCardinal(Double longitude) {
```
Comment (klezner) [inline src/main/java/com/mkgn/weatherservice/location/LocationMapper.java:21]: Unnecessary method

File: src/main/java/com/mkgn/weatherservice/location/LocationMapper.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        if (longitude > 0) {
+            return Cardinals.EAST.getCardinalAbbreviation();
+        } else if (longitude < 0) {
+            return Cardinals.WEST.getCardinalAbbreviation();
+        } else {
+            return Cardinals.PRIME_MERIDIAN.getCardinalAbbreviation();
+        }
+    }
+
+    private String latitudeCardinal(Double latitude) {
```
Comment (klezner) [inline src/main/java/com/mkgn/weatherservice/location/LocationMapper.java:31]: Unnecessary method

**Review by klezner — COMMENTED** [review]
My review

File: src/main/java/com/mkgn/weatherservice/location/LocationMapper.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    NewLocationResponse mapLocationToNewLocationResponse(Location location) {
+        return NewLocationResponse.builder().
+                id(location.getId())
+                .city(location.getCity())
+                .region(location.getRegion())
+                .country(location.getCountry())
+                .latitude(location.getLatitude())
+                .longitude(location.getLongitude())
+                .latitudeCardinal(latitudeCardinal(location.getLatitude()))
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationMapper.java:16]: as @klezner said - `Cardinal` is unnecessary

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        // when
+        Throwable exception = catchThrowable(() -> locationService.createLocation("city", "region", "country", 54.0, 181.0));
+
+        // then
+        assertThat(exception).isInstanceOf(CardinalsOutOfRangeException.class);
+        verify(locationRepository, times(0)).save(any());
+    }
+
+
+    //TODO how to test empty or null region?
```
Comment (MisQl) [inline src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java:130]: implement or remove this comment

File: src/main/java/com/mkgn/weatherservice/location/LocationService.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+    private final double NORTH_LIMIT = 90.0;
+    private final double SOUTH_LIMIT = -90.0;
+    private final double EAST_LIMIT = 180.0;
+    private final double WEST_LIMIT = -180.0;
+
+    private final LocationRepository locationRepository;
+
+    Location createLocation(String city, String region, String country, Double lat, Double lon) {
+        if (city == null || city.isBlank()) {
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationService.java:20]: consider using `commons-lang3` library and `StringUtils.isBlank()` method

http://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html
https://mvnrepository.com/artifact/org.apache.commons/commons-lang3

**Comment by MisQl** [comment]
add an integration test

File: src/main/java/com/mkgn/weatherservice/location/Cardinals.java (hunk 1 above)
Comment (MichalKulygin) [reply src/main/java/com/mkgn/weatherservice/location/Cardinals.java:3]: I use them in NewLocationResponse

File: src/main/java/com/mkgn/weatherservice/location/LocationService.java (hunk 8 above)
Comment (MichalKulygin) [reply src/main/java/com/mkgn/weatherservice/location/LocationService.java:20]: done

File: src/main/java/com/mkgn/weatherservice/location/LocationService.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+            throw new CardinalsOutOfRangeException("cardinals out of range! latitude to low");
+        }
+        if (lon > EAST_LIMIT) {
+            throw new CardinalsOutOfRangeException("cardinals out of range! longitude to high");
+        }
+        if (lon < WEST_LIMIT) {
+            throw new CardinalsOutOfRangeException("cardinals out of range! longitude to low");
+        }
+
+        if (region.equals(" ")) {
```
[the remark this answers is not in the document]
Comment (MichalKulygin) [reply src/main/java/com/mkgn/weatherservice/location/LocationService.java:39]: done

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java (hunk 7 above)
Comment (MichalKulygin) [reply src/test/java/com/mkgn/weatherservice/location/LocationServiceTest.java]: done in LocationServiceIntegrationTest.class

File: src/main/java/com/mkgn/weatherservice/location/CreateLocationRequest.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/CreateLocationRequest.java:11]: use only necessary annotations

File: src/main/java/com/mkgn/weatherservice/location/Location.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+import lombok.NoArgsConstructor;
+import org.hibernate.annotations.GenericGenerator;
+
+import javax.persistence.*;
+
+@Entity
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/Location.java:15]: use only necessary annotations

File: src/main/java/com/mkgn/weatherservice/location/LocationController.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+        String region = request.getRegion();
+        String country = request.getCountry();
+        Double latitude = request.getLatitude();
+        Double longitude = request.getLongitude();
+
+        Location location = locationService.createLocation(city, region, country, latitude, longitude);
+
+        NewLocationResponse responseBody = locationMapper.mapLocationToNewLocationResponse(location);
+
+        return ResponseEntity.status(201).body(responseBody);
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationController.java:28]:
```
ResponseEntity.status(201)
    .body(locationMapper.mapLocationToNewLocationResponse(locationsService, location));
````

File: src/main/java/com/mkgn/weatherservice/location/LocationMapper.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package com.mkgn.weatherservice.location;
+
+import org.springframework.stereotype.Component;
+
+@Component
+class LocationMapper {
+
+    NewLocationResponse mapLocationToNewLocationResponse(Location location) {
+        return NewLocationResponse.builder().
+                id(location.getId())
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationMapper.java:10]: `.id`

File: src/main/java/com/mkgn/weatherservice/location/LocationService.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+    private final double NORTH_LIMIT = 90.0;
+    private final double SOUTH_LIMIT = -90.0;
+    private final double EAST_LIMIT = 180.0;
+    private final double WEST_LIMIT = -180.0;
+
+    private final LocationRepository locationRepository;
+
+    Location createLocation(String city, String region, String country, Double latitude, Double longitude) {
+
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationService.java:21]: unecesary new line

File: src/main/java/com/mkgn/weatherservice/location/LocationService.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+        validateLatitude(latitude);
+        validateLongitude(longitude);
+
+        if (StringUtils.isBlank(region)) {
+            region = null;
+        }
+
+        Location location = new Location();
+        location.setCity(city);
+        location.setRegion(region);
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationService.java:33]: Optional.ofNullable(region).filter(Predicate.not(String::isBlank)).ifPresent(location::setRegion)

File: src/main/java/com/mkgn/weatherservice/location/LocationService.java [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    Location createLocation(String city, String region, String country, Double latitude, Double longitude) {
+
+        validateCity(city);
+        validateCountry(country);
+        validateLatitude(latitude);
+        validateLongitude(longitude);
+
+        if (StringUtils.isBlank(region)) {
+            region = null;
+        }
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationService.java:29]: get rid of it

File: src/main/java/com/mkgn/weatherservice/location/LocationService.java [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+            if (latitude > NORTH_LIMIT) {
+                throw new CardinalsOutOfRangeException("cardinals out of range! latitude to high");
+            } else {
+                throw new CardinalsOutOfRangeException("cardinals out of range! latitude to low");
+            }
+        }
+    }
+
+    private void validateLongitude(Double longitude) {
+
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationService.java:63]: unnecessary new line

File: src/main/java/com/mkgn/weatherservice/location/LocationService.java [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+    private void validateCountry(String country) {
+        if (StringUtils.isBlank(country)) {
+            throw new EmptyInputException("City cannot be empty or null");
+        }
+    }
+
+    private void validateLatitude(Double latitude) {
+        if (latitude < SOUTH_LIMIT || latitude > NORTH_LIMIT) {
+            if (latitude > NORTH_LIMIT) {
+                throw new CardinalsOutOfRangeException("cardinals out of range! latitude to high");
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationService.java:55]: what is the proper value - range?

File: src/main/java/com/mkgn/weatherservice/location/LocationService.java [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+                throw new CardinalsOutOfRangeException("cardinals out of range! latitude to low");
+            }
+        }
+    }
+
+    private void validateLongitude(Double longitude) {
+
+        if (longitude > EAST_LIMIT || longitude < WEST_LIMIT) {
+            if (longitude > EAST_LIMIT) {
+                throw new CardinalsOutOfRangeException("cardinals out of range! longitude to high");
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/LocationService.java:66]: what is the proper value - range?

File: src/main/java/com/mkgn/weatherservice/location/NewLocationResponse.java [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
```
Comment (MisQl) [inline src/main/java/com/mkgn/weatherservice/location/NewLocationResponse.java:11]: use only necessary annotations

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceIntegrationTest.java [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@SpringBootTest
+@AutoConfigureMockMvc
+class LocationServiceIntegrationTest {
+
+    @Autowired
+    MockMvc mockMvc;
+
```
Comment (MisQl) [inline src/test/java/com/mkgn/weatherservice/location/LocationServiceIntegrationTest.java:22]: unnecessary new line

File: src/test/java/com/mkgn/weatherservice/location/LocationServiceIntegrationTest.java [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+            assertThat(entry.getRegion()).isEqualTo("region");
+            assertThat(entry.getLatitude()).isEqualTo(54.0);
+            assertThat(entry.getLongitude()).isEqualTo(18.0);
+        });
+    }
+
+    @Test
+    void createLocation_whenCityIsEmpty_returns400StatusCode() throws Exception {
+        // given
+        locationRepository.deleteAll();
```
Comment (MisQl) [inline src/test/java/com/mkgn/weatherservice/location/LocationServiceIntegrationTest.java:60]:
```
   @BeforeEach
    void setUp() {
        locationsRepository.deleteAll();
    }
```

**Pull request closed without merging**